.github/workflows/build-image.yaml

```diff
@@ -0,0 +1,121 @@
+        required: true
+        type: string
+      image:
+        description: 'Name of the image'
+        required: true
+        type: string
+      image_tag:
+        description: 'Tag of the image'
+        required: true
+        type: string
+      push:
+        description: 'Push the image to the registry'
+        required: false
+        type: boolean
+        default: true
+      platforms:
+        description: 'Platforms to build for'
+        required: false
+        default: 'linux/amd64'
+        type: string
+      runs-on:
+        description: 'The type of machine to run the job on'
+        required: false
+        default: 'ubuntu-latest'
+        type: string
+    secrets:
+      username: 
+        description: 'Docker registry username'
+        required: false
+      password:
+        description: 'Docker registry token (password)'
+        required: false
+
+jobs:
+  docker:
+    runs-on: ${{ inputs.runs-on }}
+    permissions:
+      contents: read
+      packages: write
+    steps:
+      - name: Print inputs
+        run: |
+          echo "dockerfile: ${{ inputs.dockerfile }}"
+          echo "registry: ${{ inputs.registry }}"
+          echo "image: ${{ inputs.image }}"
+          echo "image_tag: ${{ inputs.image_tag }}"
+          echo "push: ${{ inputs.push }}"
+          echo "platforms: ${{ inputs.platforms }}"
+          
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          submodules: recursive
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Docker registry login
+        if: ${{ inputs.push }}
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ inputs.registry }}
+          username: ${{ secrets.username }}
+          password: ${{ secrets.password }}
+
+      - name: Set lower case image name
+        run: |
+          echo "IMAGE_LC=${IMAGE,,}" >>${GITHUB_ENV}
+        env:
+          IMAGE: ${{ inputs.registry }}/${{ inputs.image }}
+      
+      - name: Set build tag
+        id: build_tag_generator
+        run: |
+          echo "BUILD_TAG=$(date +"%Y%m%d")-$GITHUB_RUN_NUMBER" >> $GITHUB_OUTPUT
+          echo "BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ")" >> $GITHUB_OUTPUT
+
+      - name: Build and push
+        uses: docker/build-push-action@v6
+        with:
+          provenance: false
+          file: ${{ inputs.dockerfile }}
+          builder: ${{ steps.buildx.outputs.name }}
+          push: ${{ inputs.push }}
+          platforms: ${{ inputs.platforms }}
+          tags: "${{ env.IMAGE_LC }}:${{ inputs.image_tag }}"
+          labels: |
+            commit=${{ github.sha }}
+            build_date=${{ steps.build_tag_generator.outputs.BUILD_DATE }}
+            tag=${{ steps.build_tag_generator.outputs.BUILD_TAG }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
```

Comment:
Wait what happened to the images here?

Reply:
Becasue that not all images are pushed, and the images that are pushed can be pushed to different registries, we are using the gha cache mecahnisme and not the registry cache mecahnisme